# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: layout/base/PresShell.cpp
# Commit: 9aa1ce4ed2cc
# Full Hash: 9aa1ce4ed2cc25a5aaa1767e6ab63028ba24f13d
# Author: Masayuki Nakano <masayuki@d-toybox.com>
# Date: 2024-03-15 09:43:32
# Description:
#   Bug 1885259 - Make `AutoPointerEventTargetUpdater` work without event target frame r=edgar,dom-core
#   
#   The crash occurs when `PointerEventHandler::DispatchPointerFromMouseOrTouch`
#   dispatches a pointer event for `eTouchStart`.  In this case, `aFrame` of
#   the constructor of `AutoPointerEventTargetUpdater` is set to the primary frame
# ==============================================================================

diff -r cec72c3f4e5b -r 9aa1ce4ed2cc layout/base/PresShell.cpp
--- a/layout/base/PresShell.cpp	Thu Mar 14 23:40:49 2024 +0000
+++ b/layout/base/PresShell.cpp	Thu Mar 14 23:46:07 2024 +0000
@@ -565,41 +565,43 @@
 class MOZ_STACK_CLASS AutoPointerEventTargetUpdater final {
  public:
   AutoPointerEventTargetUpdater(PresShell* aShell, WidgetEvent* aEvent,
-                                nsIFrame* aFrame, nsIContent** aTargetContent) {
+                                nsIFrame* aFrame, nsIContent* aTargetContent,
+                                nsIContent** aOutTargetContent) {
     MOZ_ASSERT(aEvent);
-    if (!aTargetContent || aEvent->mClass != ePointerEventClass) {
+    if (!aOutTargetContent || aEvent->mClass != ePointerEventClass) {
       // Make the destructor happy.
-      mTargetContent = nullptr;
+      mOutTargetContent = nullptr;
       return;
     }
     MOZ_ASSERT(aShell);
-    MOZ_ASSERT(aFrame);
-    MOZ_ASSERT(!aFrame->GetContent() ||
-               aShell->GetDocument() == aFrame->GetContent()->OwnerDoc());
+    MOZ_ASSERT_IF(aFrame && aFrame->GetContent(),
+                  aShell->GetDocument() == aFrame->GetContent()->OwnerDoc());
 
     mShell = aShell;
     mWeakFrame = aFrame;
-    mTargetContent = aTargetContent;
+    mOutTargetContent = aOutTargetContent;
     mFromTouch = aEvent->AsPointerEvent()->mFromTouchEvent;
+    // Touch event target may have no frame, e.g., removed from the DOM
+    MOZ_ASSERT_IF(!mFromTouch, aFrame);
     mOriginalPointerEventTarget = aShell->mPointerEventTarget =
-        aFrame->GetContent();
+        aFrame ? aFrame->GetContent() : aTargetContent;
   }
 
   ~AutoPointerEventTargetUpdater() {
-    if (!mTargetContent || !mShell || mWeakFrame.IsAlive()) {
+    if (!mOutTargetContent || !mShell || mWeakFrame.IsAlive()) {
       return;
     }
     if (mFromTouch) {
       // If the source event is a touch event, the touch event target should
       // always be same target as preceding ePointerDown.  Therefore, we should
       // always set it back to the original event target.
-      mOriginalPointerEventTarget.swap(*mTargetContent);
+      mOriginalPointerEventTarget.swap(*mOutTargetContent);
     } else {
       // If the source event is not a touch event (must be a mouse event in
       // this case), the event should be fired on the closest inclusive ancestor
       // of the pointer event target which is still connected.  The mutations
       // are tracked by PresShell::ContentRemoved.  Therefore, we should set it.
-      mShell->mPointerEventTarget.swap(*mTargetContent);
+      mShell->mPointerEventTarget.swap(*mOutTargetContent);
     }
   }
 
@@ -607,7 +609,7 @@
   RefPtr<PresShell> mShell;
   nsCOMPtr<nsIContent> mOriginalPointerEventTarget;
   AutoWeakFrame mWeakFrame;
-  nsIContent** mTargetContent;
+  nsIContent** mOutTargetContent;
   bool mFromTouch = false;
 };
 
@@ -8339,7 +8341,7 @@
     mPresShell->RecordPointerLocation(aEvent->AsMouseEvent());
   }
   AutoPointerEventTargetUpdater updater(mPresShell, aEvent, aNewEventFrame,
-                                        aTargetContent);
+                                        aNewEventContent, aTargetContent);
   AutoCurrentEventInfoSetter eventInfoSetter(*this, aNewEventFrame,
                                              aNewEventContent);
   nsresult rv = HandleEventWithCurrentEventInfo(aEvent, aEventStatus, false,
